### ISSUE TITLE
Expressions with functions are not goroutine-safe

### ISSUE THREAD

**Issue by a13xb** [antchfx/xpath#43]
I'm using this library to batch parse a large number of HTML documents in parallel (via `antchfx/htmlquery`) and it will always result either in a nil pointer dereference or wrong result.

The issue is reuse of query objects due to shallow cloning of `functionQuery` objects.

1. [`htmlquery` calls `xpath.Expr.Select`](https://github.com/antchfx/xpath/blob/8f6e57f257db5be1c5da9fc2ee699b9186249a15/xpath.go#L129), which creates a new node iterator containing a cloned query.
2. The query is cloned recursively.
3. Eventually [`functionQuery` is cloned](https://github.com/antchfx/xpath/blob/8f6e57f257db5be1c5da9fc2ee699b9186249a15/query.go#L610) but the clone is shallow, because the function is actually a closure that [captured references to the original query objects when it was constructed](https://github.com/antchfx/xpath/blob/8f6e57f257db5be1c5da9fc2ee699b9186249a15/build.go#L186).
4. At this point the all cloned queries will reference the same original child query nodes that will be mutated concurrently when `Select` is called (for example [`attributeQuery.iterator` in `attributeQuery.Select`](https://github.com/antchfx/xpath/blob/8f6e57f257db5be1c5da9fc2ee699b9186249a15/query.go#L125)).

Here is a self-contained example with `htmlquery`, which uses a global cache of `xpath.Expr` objects, that hits the problem fairly quickly on my machine.

```go
package main

import (
	"github.com/antchfx/htmlquery"
	"golang.org/x/net/html"
	"log"
	"strings"
)

func work() {
	doc, err := html.Parse(strings.NewReader(`
		<html>
			<head><title>Hello</title></head>
			<body>
				<div class="c1 c2">A</div>
				<div class="c2 c3">B</div>
				<div class="c3 c1">C</div>
			</body>
		</html>
	`))
	if err != nil {
		log.Fatal(err)
	}
	ns := htmlquery.Find(doc, `//div[contains(@class,"c1")]/text()`)
	if len(ns) == 0 {
		log.Fatal("text not found")
	}
	elems := []string{}
	for _, n := range ns {
		elems = append(elems, htmlquery.InnerText(n))
	}
	log.Printf(`%d elems: %v`, len(elems), elems)
}

func main() {
	for i := 0; i < 4; i++ {
		go func() {
			for {
				work()
			}
		}()
	}
	select {}
}
```
Note that it will run correctly if you remove the `contains` clause, and just use `=`, for example.

My current workaround is to disable query cache in `htmlquery`, but what really needs to be done is to refactor `functionQuery` so it can be cloned correctly.

**Comment by zhengchun** [comment]
Thanks for your feedback. I added a new method to decide whether clone input argument. This might help.

**Comment by zhengchun** [comment]
Closed, fixed in `v1.1.4`.
